sixtyfps_compiler/passes/embed_images.rs

```diff
@@ -72,3 +59,216 @@ fn embed_images_from_expression(
+    }
+    let is_column_transparent = |x| {
+        for y in top..=bottom {
+            if !image.get_pixel(x, y).is_transparent() {
+                return false;
+            }
+        }
+        true
+    };
+    let mut left = 0;
+    while is_column_transparent(left) {
+        left += 1;
+        assert!(left < image.width()); // otherwise we would have a transparent image
+    }
+    let mut right = image.width() - 1;
+    while is_column_transparent(right) {
+        right -= 1;
+        assert!(right > left); // otherwise we would have a transparent image
+    }
+    let mut is_opaque = true;
+    enum ColorState {
+        Unset,
+        Different,
+        RGB([u8; 3]),
+    }
+    let mut color = ColorState::Unset;
+    'outer: for y in top..=bottom {
+        for x in left..=right {
+            let p = image.get_pixel(x, y);
+            let alpha = p[3];
+            if alpha != 255 {
+                is_opaque = false;
+            }
+            match color {
+                ColorState::Unset => {
+                    color = ColorState::RGB(p.0[0..3].try_into().unwrap());
+                }
+                ColorState::Different => {
+                    if !is_opaque {
+                        break 'outer;
+                    }
+                }
+                ColorState::RGB([a, b, c]) => {
+                    let abs_diff = |t, u| {
+                        if t < u {
+                            u - t
+                        } else {
+                            t - u
+                        }
+                    };
+                    if abs_diff(a, p[0]) > 2 || abs_diff(b, p[1]) > 2 || abs_diff(c, p[2]) > 2 {
+                        color = ColorState::Different
+                    }
+                }
+            }
+        }
+    }
+
+    let format = if let ColorState::RGB(c) = color {
+        PixelFormat::AlphaMap(c)
+    } else if is_opaque {
+        PixelFormat::Rgb
+    } else {
+        PixelFormat::Rgba
+    };
+
+    let rect = Rect::from_ltrb(left as _, top as _, (right + 1) as _, (bottom + 1) as _).unwrap();
+    Texture {
+        total_size: Size { width: image.width(), height: image.height() },
+        rect,
+        data: convert_image(image, format, rect),
+        format,
+    }
+}
+
+fn convert_image(image: image::RgbaImage, format: PixelFormat, rect: Rect) -> Vec<u8> {
+    let i = image::SubImage::new(&image, rect.x() as _, rect.y() as _, rect.width(), rect.height());
+    match format {
+        PixelFormat::Rgb => {
+            i.pixels().flat_map(|(_, _, p)| IntoIterator::into_iter(p.0).take(3)).collect()
+        }
+        PixelFormat::Rgba => {
+            i.pixels().flat_map(|(_, _, p)| IntoIterator::into_iter(p.0)).collect()
+        }
+        PixelFormat::AlphaMap(_) => i.pixels().map(|(_, _, p)| p[3]).collect(),
+    }
+}
+
+fn load_image(file: crate::fileaccess::VirtualFile) -> image::ImageResult<image::RgbaImage> {
```

Comment:
If the markup specifies a fixed width (or height) then it would probably be worthwhile to take that into account here. But I understand that this may also require knowledge about the device pixel ratio. Perhaps worth a TODO comment here?